client/src/game/reducers/cardDeductionReducer.ts

```diff
@@ -0,0 +1,343 @@
+    case "draw": {
+      return makeDeductions(deck, oldDeck, hands, metadata);
+    }
+    default: {
+      return Array.from(deck);
+    }
+  }
+}
+
+function makeDeductions(
+  deck: readonly CardState[],
+  oldDeck: readonly CardState[],
+  hands: ReadonlyArray<readonly number[]>,
+  metadata: GameMetadata,
+) {
+  const newDeck: CardState[] = Array.from(deck);
+  const variant = getVariant(metadata.options.variantName);
+  const cardCountMap = getCardCountMap(variant);
+
+  // We need to calculate our own unknown cards first because those possibilities will be needed for
+  // pretending like we know what the other players see.
+  calculatePlayerPossibilities(
+    metadata.ourPlayerIndex,
+    metadata.ourPlayerIndex,
+    hands,
+    newDeck,
+    oldDeck,
+    cardCountMap,
+  );
+  for (let playerIndex = 0; playerIndex < hands.length; playerIndex++) {
+    if (playerIndex !== metadata.ourPlayerIndex) {
+      calculatePlayerPossibilities(
+        playerIndex,
+        metadata.ourPlayerIndex,
+        hands,
+        newDeck,
+        oldDeck,
+        cardCountMap,
+      );
+    }
+  }
+  return newDeck;
+}
+
+function calculatePlayerPossibilities(
+  playerIndex: number,
+  ourPlayerIndex: number,
+  hands: ReadonlyArray<readonly number[]>,
+  deck: CardState[],
+  oldDeck: readonly CardState[],
+  cardCountMap: readonly number[][],
+) {
+  const cardsToCalculate = getCardsToCalculate(
+    playerIndex,
+    ourPlayerIndex,
+    hands,
+    deck,
+    oldDeck,
+  );
+
+  cardsToCalculate.forEach((card) => {
+    const possibilities = generatePossibilitiesForUnknownCards(
+      card.order,
+      deck,
+      playerIndex,
+      ourPlayerIndex,
+    );
+    let { possibleCards, possibleCardsForEmpathy } = card;
+    if (playerIndex === ourPlayerIndex) {
+      possibleCards = possibleCards.filter((possibility) =>
+        possibilityValid(possibility, possibilities, 0, cardCountMap),
+      );
+      if (playerIndex === card.location) {
+        // If the card is in our own hand then we also need to update empathy
+        if (card.revealedToPlayer[playerIndex]) {
+          // If it's revealed to us then we can safely just use possibleCards and don't
+          // need to do a separate calculation run for possibleCardsForEmpathy.
+          possibleCardsForEmpathy = possibleCards;
+        } else {
+          // This could be a replay where all cards are known but if it's not supposed to be
+          // revealed to us when we were in the game then we still want to be able to show some
+          // empathy.
+          possibleCardsForEmpathy = possibleCardsForEmpathy.filter(
+            (possibility) =>
+              possibilityValid(possibility, possibilities, 0, cardCountMap),
+          );
+        }
+      }
+    } else if (playerIndex === card.location) {
+      possibleCardsForEmpathy = possibleCardsForEmpathy.filter((possibility) =>
+        possibilityValid(possibility, possibilities, 0, cardCountMap),
+      );
+    }
```

Comment:
I find the logic in these ifs a bit hard to read/understand. Maybe it can be simplified to something like this?
```suggestion
    if (playerIndex === ourPlayerIndex) {
      possibleCards = possibleCards.filter((possibility) =>
        possibilityValid(possibility, possibilities, 0, cardCountMap),
      );
    }
    if (playerIndex === card.location) {
      possibleCardsForEmpathy = possibleCardsForEmpathy.filter((possibility) =>
        possibilityValid(possibility, possibilities, 0, cardCountMap),
      );
    }
```

Reply:
should be else if

Reply:
My suggestion should be else if? No. We still need to update `possibleCardsForEmpathy` on our own cards

Reply:
o

Reply:
Done. Some details are in a comment above about why it was (erroneously) like that previously.